### DISCUSSION TITLE
Declaratively define GitHub actions

### DISCUSSION THREAD

**Discussion by sidvishnoi** [community/community#26424] [Actions]
<p><strong>tl;dr:</strong> Define a GitHub Action using the workflow syntax, allowing reuse of other actions across repositories without creating own JS based wrapper actions.</p>
<p>Presently, there are two ways to define GitHub actions - either using JavaScript or Docker. If we want to reuse some action, we need to clone/install it and then make our own wrapper around it. It becomes more difficult when the action doesn’t provide a programmatic API. The workflow syntax already allows (re)use of various actions, but then one needs to copy same workflow across multiple projects. It would be great if we can define new GitHub actions by using the workflow syntax.</p>
<p>In JavaScript, we may write an action as:</p>
<pre><code class="lang-auto">// Skipping some syntax for brevity.
// ASSUMING all actions provide a programmatic API.
// All action libraries are in node_modules, checked-in to the git repository (or packaged using some other tool).
import core from "@actions/core";
import checkout from "@actions/checkout";
import exec from "some-exec-lib";
import deployToGhPages from "some-deploy-lib";
import postComment from "some-comment-lib";

async function main() {
  await checkout();
  await exec(`npm run build -- --profile ${core.getInput("some-input")}`);
  const outputs = await deployToGhPages({ token: core.getInput("github_token") });
  await postComment({ files: outputs.changed_files });
  // above gets out of hands quick.
}

main().catch((error) =&gt; {
  console.error(error);
  process.exit(1);
}
</code></pre>
<p>vs an <code>action.yml</code> in <code>my-user/declarative-action</code> may look like:</p>
<pre><code class="lang-yaml">name: 'Build, deploy and add comment'
inputs:
  SOME_INPUT:
    description: 'input description'
    required: true
steps:
  - uses: actions/checkout@v2
  - run: npm run build
    with:
      profile: ${{ inputs.SOME_INPUT }}
  - id: deploy
    uses: peaceiris/actions-gh-pages@v3
    with:
      github_token: ${{ secrets.GITHUB_TOKEN }}
  - name: comment
    uses: ...
    with:
      files: ${{ outputs.deploy.changed_files }}
</code></pre>
<p>Now copying (and updating) above in a workflow across repositories will be cumbersome. So, users can use above action in a workflow as:</p>
<pre><code class="lang-yaml">jobs:
  some_job:
    runs-on: ubuntu-latest
    steps:
      - uses: user/my-declarative-action@v1
        with:
          SOME_INPUT: my_input_value    
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/sidvishnoi">@sidvishnoi</a>,</p>
<p>Thanks for your feedback.<br>
Currently, we can’t reuse an action in another action like as the method you mentioned.</p>
<p>GitHub takes your suggestions very seriously, and the suggestions are very helpful for improving GitHub Actions.<br>
I recommend that you can directly share your suggestion <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a>. That will allow you to directly interact with the appropriate engineering team, and make it more convenient for the engineering team to collect and categorize your suggestions.</p>

**Comment by sidvishnoi** [comment]
<p>Thanks. Shared through the feedback channel you provided.<br>
Keeping this open here so community members can vote and show interest.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/sidvishnoi">@sidvishnoi</a>,</p>
<p>Sure. Thanks for your suggestions.</p>

**Comment by thboop** [comment, marked as answer]
<p>Hey <a class="mention" href="/sidvishnoi">@sidvishnoi</a>,</p>
<p>If I’m understanding your comments correctly, you are wanting a way to write an action using the yaml syntax. If so, I’m happy to report its on our radar, though I’m unable to give a timeframe for when it will be available.</p>
<p>We have an ADR open to build this on the actions runner <a href="https://github.com/actions/runner/pull/554/files" rel="nofollow noopener">here</a>. To begin, it will only support <code>run</code> steps, we will later expand it to include <code>uses</code> steps as well.</p>
<p>We are more than happy to take feedback and questions on the ADR if you have additional feedback!</p>

**Comment by sidvishnoi** [comment]
<blockquote>
<p>you are wanting a way to write an action using the yaml syntax</p>
</blockquote>
<p>Yes!</p>
<blockquote>
<p>We have an ADR open to build this on the actions runner</p>
</blockquote>
<p>This is exciting :tada:  Thank you!</p>

**Comment by softprops** [comment]
<blockquote>
<p>If I’m understanding your comments correctly, you are wanting a way to write an action using the yaml syntax.</p>
</blockquote>
<p>This seems possible today using an action like actions/github-script which allows you to inline javascript . You can find an example here <a href="https://github.com/actions/github-script#comment-on-an-issue" rel="nofollow noopener">https://github.com/actions/github-script#comment-on-an-issue</a></p>

**Comment by Simran-B** [comment]
<p>Composite run steps actions are also available since August:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action" target="_blank" rel="noopener">Creating a composite run steps action - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by Riapernecita** [comment]
Yes you will interact them but specific about your self dont give them a number that using a bank and dont save your password to apps and dont used automatic signed in always keep it in your mind ..always used a specific number for using a like gaming, your social media and dont forget your service developer trust them , we have a github policy..

**Comment by sidvishnoi** [comment]
Composite actions have resolved this issue (long ago).
